### PR TITLE
Healthchecking

### PR DESCRIPTION
**Pull request by enxebre**
- initial support for machine healthchecking:
-- Revendor for operator sdk and so controller runtime
-- Introduce machineHealthChecking API
-- Introduce machineHealthCheck controller for watching nodes and get backing machines belonging to an existing healthcheck
- Follow up:
-- Remediation logic

### CODE REVIEW

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+				},
+				Status: healthcheckingv1alpha1.MachineHealthCheckStatus{},
+			},
+			expected: false,
+		},
+	}
+
+	for _, tc := range testsCases {
+		if got := hasMatchingLabels(tc.machineHealthCheck, tc.machine); got != tc.expected {
+			t.Errorf("Expected %t, got %t", tc.expected, got)
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Is this meant to be `%T`, or `%v`?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		}
+	}
+}
+
+func TestIsHealthy(t *testing.T) {
+	nodeHealthy := node("healthy", true)
+	nodeUnhealthy := node("unhealthy", false)
+
+	if got := isHealthy(nodeHealthy); got != true {
+		t.Errorf("Expected true, got %t", got)
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Is this meant to be `%T`, or `%v`?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+	}
+	if got := isHealthy(nodeUnhealthy); got != false {
+		t.Errorf("Expected false, got %t", got)
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Is this meant to be `%T`, or `%v`?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+			t.Errorf("Expected %t, got %t", tc.expected, got)
+		}
+	}
+}
+
+func TestIsHealthy(t *testing.T) {
+	nodeHealthy := node("healthy", true)
+	nodeUnhealthy := node("unhealthy", false)
+
+	if got := isHealthy(nodeHealthy); got != true {
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: use `; !got`

Actually, `!healthy` would read better.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+		t.Errorf("Expected true, got %t", got)
+	}
+	if got := isHealthy(nodeUnhealthy); got != false {
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Prefer `healthy`.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+	r := newFakeReconciler(nodeHealthy, nodeUnhealthy, fakeMachine, allMachineHealthChecks)
+	for _, tc := range testsCases {
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: "",
+				Name:      tc.node.Name,
+			},
+		}
+		result, err := r.Reconcile(request)
+		if result != tc.expected.result {
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Should this not check `err` first, before asserting on `result`?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 2 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: is %t for booleans right https://golang.org/pkg/fmt/?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 6 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: not really, result and err are the tuple expectation

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+	"testing"
+)
+
+const (
+	namespace = "openshift-cluster-api"
+)
+
+func init() {
+	// Add types to scheme
+	capiv1alpha1.AddToScheme(scheme.Scheme)
```
[the remark this answers is not in the document]
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: https://github.com/kubernetes-sigs/controller-runtime/issues/137

File: cmd/machine-healthcheck/main.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+	// Setup all Controllers
+	if err := controller.AddToManager(mgr); err != nil {
+		glog.Fatal(err)
+	}
+
+	glog.Info("Starting the Cmd.")
+
+	// Start the Cmd
+	glog.Fatal(mgr.Start(signals.SetupSignalHandler()))
```
[the remark this answers is not in the document]
Comment (enxebre) [reply cmd/machine-healthcheck/main.go]: sounds ok to me here, what would you suggest otherwise
```
if err Fatal else exit(0)
```
?

**Pull request merged**